src/rust/ide/src/controller/graph.rs

```diff
@@ -0,0 +1,218 @@
+}
+
+/// Describes the desired position of the node's line in the graph's code block.
+#[derive(Clone,Copy,Debug)]
+pub enum LocationHint {
+    /// Try placing this node's line before the line described by id.
+    Before(ast::ID),
+    /// Try placing this node's line after the line described by id.
+    After(ast::ID),
+    /// Try placing this node's line at the start of the graph's code block.
+    Start,
+    /// Try placing this node's line at the end of the graph's code block.
+    End,
+}
+
+
+
+// ==================
+// === Controller ===
+// ==================
+
+/// State data of the module controller.
+#[derive(Clone,Debug)]
+pub struct Handle {
+    /// Controller of the module which this graph belongs to.
+    module    : controller::module::Handle,
+    id        : Id,
+    /// Publisher. When creating a controller, it sets up task to emit notifications through this
+    /// publisher to relay changes from the module controller.
+    publisher : Rc<RefCell<controller::notification::Publisher<controller::notification::Graph>>>,
+}
+
+impl Handle {
+    /// Gets a handle to a controller of the module that this definition belongs to.
+    pub fn get_module(&self) -> controller::module::Handle {
+        self.module.clone()
+    }
+
+    /// Gets a handle to a controller of the module that this definition belongs to.
+    pub fn get_id(&self) -> Id {
+        self.id.clone()
+    }
+
+    /// Creates a new controller. Does not check if id is valid.
+    ///
+    /// Requires global executor to spawn the events relay task.
+    pub fn new_unchecked(module:controller::module::Handle, id:Id) -> Handle {
+        let graphs_notifications = module.subscribe_graph_notifications();
+        let publisher = default();
+        let ret       = Handle {module,id,publisher};
+        let weak      = Rc::downgrade(&ret.publisher);
+        let relay_notifications = process_stream_with_handle(
+            graphs_notifications,
+            weak,
+            |notification,this| {
+                match notification {
+                    controller::notification::Graphs::Invalidate =>
+                    this.borrow_mut().publish(controller::notification::Graph::Invalidate),
+            }
+        });
+        executor::global::spawn(relay_notifications);
+        ret
+    }
+
+    /// Creates a new graph controller. Given ID should uniquely identify a definition in the
+    /// module. Fails if ID cannot be resolved.
+    ///
+    /// Requires global executor to spawn the events relay task.
+    pub fn new(module:controller::module::Handle, id:Id) -> FallibleResult<Handle> {
+        let ret = Self::new_unchecked(module,id);
+        let _ = ret.get_graph_definition_info()?; // make sure that definition exists
+        Ok(ret)
+    }
+
+    /// Retrieves double rep information about definition providing this graph.
+    pub fn get_graph_definition_info
+    (&self) -> FallibleResult<double_representation::definition::DefinitionInfo> {
```

Comment:
```suggestion
    pub fn get_graph_definition_info(&self)
    -> FallibleResult<double_representation::definition::DefinitionInfo> {
```
Isn't that better?

Reply:
It is, however our guidelines state otherwise.